src/netdev.c

```diff
@@ -94,52 +94,71 @@ ni_netdev_slaveinfo_destroy(ni_slaveinfo_t *slave)
 	memset(slave, 0, sizeof(*slave));
 }
 
-static void
-ni_netdev_free(ni_netdev_t *dev)
+void
+ni_netdev_reset(ni_netdev_t *dev)
 {
-	/* Clear out linkinfo */
-	ni_string_free(&dev->link.qdisc);
-	ni_string_free(&dev->link.kind);
+	if (!dev)
+		return;
+
+	/* Clear and Reset all (kernel) property data,
+	 * except of list, ref counter & marker flags */
+	ni_string_free(&dev->name);
+	dev->link.type = NI_IFTYPE_UNKNOWN;
+	dev->link.ifindex = 0;
+	dev->link.ifflags = 0;
+
+	ni_link_address_init(&dev->link.hwaddr);
+	ni_link_address_init(&dev->link.hwpeer);
 	ni_string_free(&dev->link.alias);
+
+	dev->link.mtu = 0;
+	dev->link.metric = 0;
+	dev->link.txqlen = 0;
+	dev->link.oper_state = 0;
+
 	ni_netdev_ref_destroy(&dev->link.lowerdev);
 	ni_netdev_ref_destroy(&dev->link.masterdev);
 	ni_netdev_slaveinfo_destroy(&dev->link.slave);
+
+	ni_string_free(&dev->link.qdisc);
+	ni_string_free(&dev->link.kind);
+
 	ni_netdev_set_link_stats(dev, NULL);
+	ni_netdev_set_client_state(dev, NULL);
 
-	/* Clear out addresses, routes, ... */
 	ni_netdev_clear_addresses(dev);
 	ni_netdev_clear_routes(dev);
-	ni_netdev_set_ethernet(dev, NULL);
-	ni_netdev_set_infiniband(dev, NULL);
-	ni_netdev_set_bonding(dev, NULL);
+
+	ni_netdev_set_ipv4(dev, NULL);
+	ni_netdev_set_ipv6(dev, NULL);
+	ni_netdev_set_auto6(dev, NULL);
```

Comment:
Side note:
https://github.com/openSUSE/wicked/blob/468a60bcf107b7375702490a4a1171860429d852/src/auto6.c#L171
never set anything, it always do a free...

Reply:
Well, true. There is no config to assign (should be `ni_auto6_conf_t *conf`) like the sysctl's in ipv4/ipv6 -- it's currently
state only always filled by kernel events.
I didn't changed any of these functions, just their call order to match order in the struct to ensure we don't miss to reset
any struct member (and cause a leak). When, then it is a separate thing to cleanup next time we touch auto6.